paasta_tools/contrib/paasta_update_soa_memcpu.py

```diff
@@ -165,29 +170,83 @@ def get_report_from_splunk(creds, app, filename, criteria_filter):
-        serv["hacheck_cpus"] = d["result"].get("suggested_hacheck_cpus")
-        serv["cpu_burst_add"] = d["result"].get("suggested_cpu_burst_add")
-        services_to_update[criteria] = serv
+
+        # fill in fields that don't need any transformation
+        serv = {
+            "service": criteria.split(" ")[0],
+            "cluster": criteria.split(" ")[1],
+            "instance": criteria.split(" ")[2],
+            "owner": d["result"].get("service_owner", "Unavailable"),
+            "date": d["result"]["_time"].split(" ")[0],
+            "money": d["result"].get("estimated_monthly_savings", 0),
+            "project": d["result"].get("project", "Unavailable"),
+        }
+
+        # the report we get is all strings, so we need to convert them to the right types
+        field_conversions = {
+            # the general format here is:
+            # expected_key: (source_key, conversion_function)
+            # where expected_key is the key we want in the final output and source_key is the key from the splunk
+            # report
+            "cpu_burst_add": ("suggested_cpu_burst_add", float),
+            "cpus": ("suggested_cpus", float),
+            "disk": ("suggested_disk", int),
+            "hacheck_cpus": ("suggested_hacheck_cpus", float),
+            "max_instances": ("suggested_max_instances", int),
+            "mem": ("suggested_mem", int),
+            "min_instances": ("suggested_min_instances", int),
+            "old_cpus": ("current_cpus", float),
+            # not quite sure why these are floats...they're ints in soaconfigs
+            "old_disk": ("current_disk", _force_str_to_int),
+            "old_mem": ("current_mem", _force_str_to_int),
+        }
+
+        for key, (source_key, conversion_function) in field_conversions.items():
+            value = d["result"].get(source_key)
+            serv[key] = conversion_function(value) if value is not None else None
+
+        # merge results if we've already seen rows for this service
+        # NOTE: this is necessary since the Splunk search can return multiple rows
+        # for the same (service, cluster, instance) tuple as the autotune query
+        # treats certain cpu allocation changes as if the tuple was entirely different.
+        # this is ostensibly due to a theory that if you update resource allocation, existing
+        # autotune data is potentially invalidated - but in practice this ends up hampering
+        # autotune for services with highly variable resource allocation - e.g., we have some services
+        # that have their cpu allocation tweaked by +/-.1 cpu pretty frequently, but then min/max autotune
+        # is never updated.
+        if criteria in services_to_update:
+            for key in serv:
+                # we probably don't want to merge any other fields since they're going to be strings :p
+                if key not in field_conversions:
+                    continue
```

Comment:
This might change results since we'd have been taking e.g. `*_reason` fields from only the latest row previously, but now we won't be overwriting those any longer even if we are overwtiting the values.  

I think it might add some complexity trying to make sure `_reason` fields stay with their respective values unfortunately tho; we'd probably need to consider a row as a distinct datapoint and figure out how to take the max of a datapoint + write all fields for that instead of going field by field?

Reply:
oh, although I guess the `reason` is not really relevant here, and only in our splunk-based investigation dashboard, so i suppose this is moot

Reply:
proving i can't read even further, i now see we don't even have reason fields in these `serv` dicts, apologies

Reply:
yea, all of this is mighty confusing - i'm hoping that in autotune v2 we move this out of contrib/ and add a whole lot more comments and tests :p